### PR TITLE
Implements NormStabilizer layer and adds tests

### PR DESCRIPTION
**Pull request by juesato**
Test command is `rnn.test{"NormStabilizer"}`. The test compares numerical gradient (with added regularization loss) to the analytically computed gradients using the `NormStabilizer` layer.

### CODE REVIEW

**Comment by nicholas-leonard** [comment]
@juesato Thank you very much for taking the time to do this! 

I modified it in the latest commits to make it faster and I also added some documentation.

Do you think you could send a PR to add it as an option to the recurrent-language-model.lua example script?

**Comment by juesato** [comment]
@nicholas-leonard No problem, thanks for the help throughout! I've been using this library a lot, so I'm glad to be able to help out where I can, so that you can focus on other things.

Sure, I can add it to `recurrent-language-model.lua`. I'm just running the code right now and will submit a PR later today.

**Comment by juesato** [comment]
@nicholas-leonard Just wanted to say that this might be a little while. I'm having trouble replicating the overfitting seen without norm stabilization in the original paper (page 3 http://arxiv.org/pdf/1511.08400.pdf), which was the whole point of the regularization. I've run the network for ~100 epochs and the validation loss is still decreasing even with no regularization. 

I suspect the difference is some hyperparameter setting (for example, the learning rate is done differently in the paper, so it'd require modifying the script - there may be other differences). I'll let you know what happens in a few days.

**Comment by nicholas-leonard** [comment]
@juesato Talking to the author of the paper, it seems that he was dividing all costs by `T=50`. In Torch, we omit this division by T. So whatever beta and learning rate you see in the paper should be respectively divided and multiplied by T to reproduce the experiment in Torch. So maybe try a `beta=50` becomes `beta=1` (the default) and a learning rate of `lr=0.002` becomes `lr=0.1`. Let me know if this helps. If not, we can include David Kruger (the main author) in the discussion.

**Comment by juesato** [comment]
@nicholas-leonard Sorry, I'm confused on a few points.

1. Alex Graves uses a similar LSTM for character-level modeling on Treebank and reports perplexity 167 / 1.32 bits-per-character after 9 epochs (http://arxiv.org/pdf/1308.0850v5.pdf Table 1, top of page 8). He also uses a single hidden layer with 1000 LSTM units, learning rate 0.0001 and momentum 0.99, gradient clipping at 1. In the norm stabilization paper (http://arxiv.org/pdf/1511.08400.pdf Figure 1, top of page 3), it looks like with no regularization, the model converges to about 1.5 BPC after ~50 epochs. It seems like the hyperparameters are essentially the same (1000 LSTM units, learning rate .002, momentum .99, gradient clipping 1), and if anything should converge faster due to the higher learning rate. I'm not sure how to understand this difference in number of epochs, and performance - both seem pretty significant.

2. Since the example script uses word-level language modeling, the word-level benchmarks in Graves' paper seem like the most natural comparison, which is perplexity ~138 after 11 epochs. I've been seeing perplexity ~250 after ~300 epochs (with trying to match hyperparameters) which also seems really different. 

3. Do you know what perplexities for the sample commands inside `recurrence-language-model.lua` are?

**Comment by nicholas-leonard** [comment]
@juesato For PTB, the learning rate schedule is extremely important as it otherwise overfits. For example, the following will get you about 76 PPL after 28 epochs : 
```
th rnn/examples/recurrent-language-model.lua --progress --cuda --lstm --seqlen 35 --uniform 0.04 --hiddensize '{1500,1500}' --batchsize 20 --startlr 1 --cutoff 10 --maxepoch 50 --schedule '{[15]=0.87,[16]=0.76,[17]=0.66,[18]=0.54,[19]=0.43,[20]=0.32,[21]=0.21,[22]=0.10}' --dropout 0.65
``` 
without dropout, using a different hyperparams, should yield about 116 PPL after less than 10 epochs (1h of training):
```
th rnn/examples/recurrent-language-model.lua --progress --cuda --lstm --seqlen 20 --uniform 0.1 --hiddensize '{200,200}' --batchsize 20 --startlr 1 --cutoff 5 --maxepoch 50 --schedule '{[15]=0.87,[16]=0.76,[17]=0.66,[18]=0.54,[19]=0.43,[20]=0.32,[21]=0.21,[22]=0.10}'
```

Maybe try the above commands with and without norm-stabilizer. For the dropout one, try with norm stabilizer but without dropout to see which one works best.

What commands were you running? 250 epochs seems like a lot for PTB.

**Comment by juesato** [comment]
@nicholas-leonard Sorry for the delay - had a bunch of things come up this last week.

The really slow convergence was because of the factor of 50 difference you mentioned above - I was trying to copy the paper hyperparameters, so the learning rate I was using was a factor of 50 too small.

I'm not getting quite the same numbers for some reason - the second command is giving me a perplexity of 128 (minimized after 6 epochs) rather than 116. Adding the norm stabilizer doesn't seem to change too much though perhaps it helps slightly: with `beta=1` the validation perplexity is minimized at 127 after 7 epochs (equivalent of `beta=50` in the paper), and with `beta=10` the validation perplexity is minimized at 126 after 8 epochs.

**Comment by juesato** [comment]
@nicholas-leonard Another update, the dropout model I trained reaches validation perplexity of 81 after 24 epochs. So also slightly worse performance than what you got. Is there any obvious reason these numbers might not match up?

**Comment by nicholas-leonard** [comment]
@juesato Did you run th scripts/evaluate-rnnlm.lua to get those numbers?

**Comment by juesato** [comment]
@nicholas-leonard Oh I've been using the validation numbers from after each epoch in the training script, but I'll try that

**Comment by juesato** [comment]
Using the evaluation script gives slightly lower numbers (I'm seeing 125 rather than 128 with no regularization, then 124 with `beta=1` and 122 with `beta=10). They're slightly different but I'd guess the test and validation perplexities should be pretty similar.

**Comment by nicholas-leonard** [comment]
@juesato Let me run the above command to see if I can reproduce. I will get back to you with results.

**Comment by juesato** [comment]
Sounds good, thanks!

**Comment by nicholas-leonard** [comment]
@juesato I gave you the wrong schedule for the second (non-dropout) command. Running this : 
```lua
th examples/recurrent-language-model.lua --progress --cuda --lstm --seqlen 20 --uniform 0.1 --hiddensize '{200,200}' --batchsize 20 --startlr 1 --cutoff 5 --maxepoch 50 --schedule '{[5]=0.5,[6]=0.25,[7]=0.125,[8]=0.0625,[9]=0.03125,[10]=0.015625,[11]=0.0078125,[12]=0.00390625}' --device 
```
Gives me a perplexity of 50, 120, 115 on train, valid and test sets, respectively, after 7 epochs.
